runtime/src/commitment.rs

```diff
@@ -103,6 +103,22 @@ impl BlockCommitmentCache {
         self.highest_confirmed_slot
     }
 
+    pub fn highest_gossip_confirmed_slot(&self) -> Slot {
+        // TODO: see solana_core::RpcSubscriptions:
+        //self.last_checked_slots.get(&CommitmentLevel::SingleGossip).unwrap_or(&0)
+        self.highest_confirmed_slot
+    }
+
+    pub fn slot_with_commitment(&self, commitment_level: CommitmentLevel) -> Slot {
+        match commitment_level {
+            CommitmentLevel::Recent => self.slot(),
+            CommitmentLevel::Root => self.root(),
```

Comment:
Here's the other change. This was `bank_forks.root()`, not the root at the time voting was evaluated.

Reply:
is there any reason not to use the latest root from BankForks if it's readily available?

 I could see the motivation if we were trying to avoid the `BankForks` lock (in case of a lot of RPC spam), but we're already holding the `BankForks` read lock by this point.

Reply:
I think ideally if the `BlockCommitmentCache` cached the `Arc<Bank>` for each of these options `highest_confirmed_slot`, `highest_confirmed_root`, `root`, we could probably avoid this lock() -> get() pattern

Reply:
Rpc can choose to `BankForks::root()` if this doesn't work out. This code is in BlockCommitmentCache, which doesn't hold a reference to BankForks.

I like the idea of BlockCommitmentCache holding all `Arc<Bank>` instead of slots. Then RPC wouldn't need to hold BankForks at all. Easy to imagine that reducing a lot of lock contention.

How about we move forward with this PR as-is and if it works out well, we can take that next step of switching to `Arc<Bank>`, which would be a bigger change and tricker to revert.

Reply:
I like the idea of switching to `Arc<Bank>`, but it may bring the issue of highest_confirmed_slot/root not existing in BankForks to a head, since the commitment service will always require those banks to exist in BankForks. BankForks pruning will also depend on whether BlockCommitmentCache still holds a reference to any banks.

Reply:
Yeah I think it's fine if we follow the changes suggested in our discussion here: https://discord.com/channels/428295358100013066/478692221441409024/732812908555272212. 

Because BankForks always keeps everything on the rooted path in the range `highest_confirmed_root` and its current `root`, I think it's safe to assume that any new `new_highest_confirmed_root > old_highest_confirmed_root` you discover should exist in BankForks, *unless* the cluster makes two confirmed roots that are not descended from one another.

I think the same should hold true for `highest_confirmed_slot ` if we define it to be  the highest such slot that descends from `max(highest_confirmed_root, our_current_root)`

Reply:
If I understand correctly, sounds like RPC ought to be failing to get `highest_confirmed_root` regularly, since BankForks could prune it shortly after passing that slot to BlockCommitmentCache. And the others are only available because the slot is always higher.

Reply:
The cache's current `highest_confirmed_root` is passed into BankForks on `set_root` explicitly to retain the range `highest_confirmed_root` to `root`